src/ds/ptrwrap.h

```diff
@@ -38,67 +38,138 @@ namespace snmalloc
    * *E forms are "exported" and have had platform constraints applied.  That
    * means, for example, on CheriBSD, that they have had their VMMAP permission
    * stripped.
-   *
-   * Yes, I wish the start-of-comment characters were aligned below as well.
-   * I blame clang format.
    */
-  enum capptr_bounds
+
+  namespace capptr_bounds
   {
-    /*                Spatial  Notes                                      */
-    CBArena, /*        Arena                                              */
-    CBChunkD, /*       Arena   Chunk-bounded in debug; internal use only! */
-    CBChunk, /*        Chunk                                              */
-    CBChunkE, /*       Chunk   (+ platform constraints)                   */
-    CBAlloc, /*        Alloc                                              */
-    CBAllocE /*        Alloc   (+ platform constraints)                   */
+    /*
+     * Bounds dimensions are sorted so that < reflects authority.  For example,
+     * spatial::Alloc is less capable than spatial::Arena.
+     */
+    enum class spatial
+    {
+      Alloc,
+      Chunk,
+      ChunkD,
+      Arena
+    };
+    enum class platform
+    {
+      Exported,
+      High
+    };
+    enum class wild
+    {
+      Unchecked,
+      Checked
+    };
+
+    template<spatial S, platform P, wild W>
+    struct t
+    {
+      static constexpr enum spatial spatial = S;
+      static constexpr enum platform platform = P;
+      static constexpr enum wild wild = W;
+
+      template<enum spatial SO>
+      using with_spatial = t<SO, P, W>;
+
+      template<enum platform PO>
+      using with_platform = t<S, PO, W>;
+
+      template<enum wild WO>
+      using with_wild = t<S, P, WO>;
+
+      /*
+       * The dimensions here are not used completely orthogonally.  In
+       * particular, high-authority spatial bounds imply high-authority in other
+       * dimensions and unchecked pointers must be annotated as tightly bounded.
+       */
+      static_assert(
+        !(S == spatial::Arena) || (P == platform::High && W == wild::Checked));
+      static_assert(
+        !(W == wild::Unchecked) ||
+        (S == spatial::Alloc && P == platform::Exported));
+    };
```

Comment:
So I was wondering if 

> Alloc => Exported

If the Alloc is does not cover all the pages it is part of, then we don't want to give out `High`, but if it is 4KiB aligned and sized, then it is okay to give out the ability to call `mprotect` to the client.  But the allocator would have be clear the `mprotect` on return?

Reply:
It is not currently the case that `spatial::Alloc` implies `platform::Exported`.  In particular, when we are reconstructing internal pointer(s) from the `wild::Unchecked` pointer given to `dealloc` by the client, we do not apply platform constraints before placing messages on the `Remote` queues.  We certainly could, but because `Mediumslab`s do not actually use the pointer given to them except as an address, it seemed excessive.  However, in thinking about that more, perhaps that's not true: if it's worth bounding `Remote`s to `spatial::Alloc` (to thwart UAF as a vector to access allocator metadata, e.g.), it's probably worth bounding them to `platform::Exported` as well.  Probably merits additional thought.

Reply:
It might actually make sense to have AAL-directed policies here... on CHERI+MTE, any bounding we do of `Remote` pointers or other bits of in-band metadata will be out of paranoia about bugs in `snmalloc` itself, as we will ensure that the client cannot read the band.  OTOH, for just CHERI, it probably is worth having the spatial bounds in place as a kind of lightweight, partial defence against UAF?  For non-CHERI, of course, all this is just type-level fun and games.